### ISSUE TITLE
Nest-cli depends on tree-kill, which depends on `ps` command, which may be unavailable and is not documented

### ISSUE THREAD

**Issue by OJezu** [nestjs/nest-cli#484]
## I'm submitting a...

<pre><code>
[ ] Regression 
[X] Bug report
[ ] Feature request
[X] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead post your question on Stack Overflow.
</code></pre>

## Current behavior

When application is started with `yarn run start:dev` inside a container with no `ps` program available, after first code change application crashes completely (along with the container)
```
$ yarn run start:dev
5:39:37 PM - File change detected. Starting incremental compilation...


5:39:39 PM - Found 0 errors. Watching for file changes.
events.js:186
      throw er; // Unhandled 'error' event
      ^

Error: spawn ps ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:264:19)
    at onErrorNT (internal/child_process.js:456:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
Emitted 'error' event on ChildProcess instance at:
    at Process.ChildProcess._handle.onexit (internal/child_process.js:270:12)
    at onErrorNT (internal/child_process.js:456:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  errno: 'ENOENT',
  code: 'ENOENT',
  syscall: 'spawn ps',
  path: 'ps',
  spawnargs: [ '-o', 'pid', '--no-headers', '--ppid', 185 ]
}
```

## Expected behavior

Application reloads with new code changes

## Minimal reproduction of the problem with instructions

1. Create new docker container from `node` image
2. Create new app with nest-cli inside the container
3. Start the app with `npm run start:dev`
4. Change something in application code
5. `npm run start:dev` crashes completely

## What is the motivation / use case for changing the behavior?

Unexpected crashes.

Possible fixes:
- dont' use `tree-kill`, make child a detached process with `detach` option, and use `process.kill(-childrenPid)` to kill entire process tree,  faff around with signals and stdio to make sure child process' tree is killed when needed  https://nodejs.org/api/child_process.html#child_process_options_detached
- add `ps` dependency to documentation
- check for `ps` being available and print sensible error
- ask nicely `tree-kill` to output more legible error when no `ps` is available

## Environment

```
 _   _             _      ___  _____  _____  _     _____
| \ | |           | |    |_  |/  ___|/  __ \| |   |_   _|
|  \| |  ___  ___ | |_     | |\ `--. | /  \/| |     | |
| . ` | / _ \/ __|| __|    | | `--. \| |    | |     | |
| |\  ||  __/\__ \| |_ /\__/ //\__/ /| \__/\| |_____| |_
\_| \_/ \___||___/ \__|\____/ \____/  \____/\_____/\___/


[System Information]
OS Version     : Linux 5.0
NodeJS Version : v12.10.0
YARN Version    : 1.19.2
[Nest Information]
platform-express version : 6.7.2
typeorm version          : 6.2.0
common version           : 6.7.2
core version             : 6.7.2

```

**Comment by kamilmysliwiec** [comment]
> ask nicely tree-kill to output more legible error when no ps is available

Would you like to create an issue in the `tree-kill` repo?

- add ps dependency to documentation
- check for ps being available and print sensible error

We can do both on our side ^

**Comment by b4dnewz** [comment]
I've encountered this problem right now, @kamilmysliwiec  if you are interested I can make a PR with the first proposed solution:

> dont' use tree-kill, make child a detached process with detach option

I'm running nest in dev mode inside a docker container because I'm interacting with other services and it's a features/fix that I really need

**Comment by OJezu** [comment]
@b4dnewz not sure it will work on Windows though, might still have to use `taskkill` program there.
Also, for UNIX-like, if children start their own detached children, those grandchildren won't be hunted down and killed. However this may be an expected outcome - after all they are detached from their parent, so they should not be auto-killed when parent dies.

**Comment by b4dnewz** [comment]
@OJezu good point, but I have few questions

> if children start their own detached children

it this the case for nestjs start:dev? i'm not familiar with how it's working in deep but looks like it will [spawn one per time](https://github.com/nestjs/nest-cli/blob/master/actions/start.action.ts#L83) process after the build has completed successfully

> not sure it will work on Windows though

I'm not sure either but after looking round in github issues I found this [comment](https://github.com/nodejs/node/issues/2098#issuecomment-169469019) and this [comment](https://github.com/nodejs/node/issues/2098#issuecomment-169549789) in the same page, which may lead to a nice and clean solution, I still haven't tested

**Comment by OJezu** [comment]
@b4dnewz start:dev starts the application, if that application starts it's own children is not up to nest.

The comments below mention, that killing process group (with `kill(-pid)`) works on non-windows. I don't see nothing there about killing process groups on windows.

**Comment by b4dnewz** [comment]
yep you right, I misunderstood the comment, my english understanding sucks..

**Comment by jexperton** [comment]
I just spent a couple of days investigating this issue, including unsuccessfully asking for information on Discord.

I've been working on other projects with *ts-node-dev* which does the same as *start:dev* so I knew it could work with Docker.

After digging into the nest-cli code, I've found _killProcess_ function. 

This made me realized that I should have clicked on this slightly misleading issue title when I first saw it in the issue list two days ago.

### **Don't use official node slim images. Use full images**

+1 for a better documentation.

**Comment by OJezu** [comment]
@jexperton re: the issue titile
If you were to look up the error message [using issue search](https://github.com/nestjs/nest-cli/issues?q=%22spawn+ps+ENOENT%22), you would find this issue. I'm sorry if the title is not searchable, but I'd rather have the title describe the issue, not the symptom. For searching... well, use search, GitHub's, or [Google](https://www.google.com/search?q=nest%20start:dev%20%22spawn%20ps%20ENOENT%22).

**Comment by jexperton** [comment]
@OJezu you're right I should blame my search skills not the issue title. That's why I agree with documenting this issue. It could be interesting to add a short advice such as:

> If you use Docker for development, make sure that your image has *ps* installed by running the following command:  `docker run --rm --entrypoint ps YOUR_IMAGE &> /dev/null && echo "installed" || echo "missing"`

**Comment by jmlopez-rod** [comment]
Ran into this issue today. After installing `ps` on my docker image it works:

```
FROM node:12.18.3-buster-slim

RUN apt-get update && apt-get install -y procps
```

I think a note on using Docker like the one @jexperton suggested would help.

**Comment by dimacros** [comment]
Lol, this is surely a common problem for those who use Docker.

**Comment by befabry** [comment]
Still running in this issue. Quite annoying

**Comment by robertkraig** [comment]
this solved my issue. Install `procps`

**Comment by yogeshwaran-c** [comment]
This issue appears to be resolved. The CLI now uses its own built-in \`tree-kill\` implementation (\`lib/utils/tree-kill.ts\`) that handles the missing \`ps\` command gracefully:

\`\`\`typescript
const result = spawnSync('ps', ['-A', '-o', 'pid,ppid'], {
  encoding: 'utf-8',
  stdio: 'pipe',
});

if (result.error || !result.stdout) {
  return [];  // Gracefully returns empty when ps is unavailable
}
\`\`\`

When \`ps\` is unavailable (e.g., in minimal containers), \`getAllPid()\` returns an empty array, so \`getAllChildren()\` returns no children. The target process itself is still killed via \`process.kill(pid, signal)\` — only child process cleanup is skipped. No runtime error is thrown.

On Windows, \`taskkill /pid X /T /F\` is used instead, which doesn't depend on \`ps\` at all.

This could potentially be closed.
